src/BuiltInTools/dotnet-watch/HotReload/CompilationHandler.cs

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Locator;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.ExternalAccess.DotNetWatch;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Tools.Internal;
+
+namespace Microsoft.DotNet.Watcher.Tools
+{
+    internal class CompilationHandler : IDisposable
+    {
+        private readonly IReporter _reporter;
+        private Task<(Solution, DotNetWatchEditAndContinueWorkspaceService)>? _initializeTask;
+        private Solution? _currentSolution;
+        private DotNetWatchEditAndContinueWorkspaceService? _editAndContinue;
+        private IDeltaApplier? _deltaApplier;
+
+        public CompilationHandler(IReporter reporter)
+        {
+            _reporter = reporter;
+        }
+
+        public async ValueTask InitializeAsync(DotNetWatchContext context, CancellationToken cancellationToken)
+        {
+            if (_deltaApplier is null)
+            {
+                _deltaApplier = context.DefaultLaunchSettingsProfile.HotReloadProfile == "blazorwasm" ?
+                    new BlazorWebAssemblyDeltaApplier(_reporter) :
+                    new AspNetCoreDeltaApplier(_reporter);
+            }
+
+            await _deltaApplier.InitializeAsync(context, cancellationToken);
+
+            if (context.Iteration == 0)
+            {
+                var instance = MSBuildLocator.QueryVisualStudioInstances().First();
+
+                _reporter.Verbose($"Using MSBuild at '{instance.MSBuildPath}' to load projects.");
+                MSBuildLocator.RegisterInstance(instance);
+            }
+            else if (_currentSolution is not null)
+            {
+                _currentSolution.Workspace.Dispose();
+                _currentSolution = null;
+            }
+
+            _initializeTask = Task.Run(() => CompilationWorkspaceProvider.CreateWorkspaceAsync(context.FileSet.Project.ProjectPath, _reporter, cancellationToken), cancellationToken);
+
+            return;
+        }
+
+        public async ValueTask<bool> TryHandleFileChange(DotNetWatchContext context, FileItem file, CancellationToken cancellationToken)
+        {
+            if (!file.FilePath.EndsWith(".cs", StringComparison.Ordinal) &&
```

Comment:
Do we need to check `.resx` or `.csx` here?

Reply:
Idk what csx is, but thiis specifically includes files that can be compiled. resx currently relies on msbuild tasks. Also, we don't watch for those by default.

Reply:
csx is a c# script file. They aren't common. https://stackoverflow.com/questions/25405941/what-are-csx-c-sharp-files-for. I don't even see a microsoft doc for them.

Reply:
Yup, definitely not trying to watch them. Plus they won't work with enc